README.md
# GQL

[![Package](https://img.shields.io/hexpm/v/gql.svg)](https://hex.pm/packages/gql) [![Documentation](http://img.shields.io/badge/hex.pm-docs-green.svg?style=flat)](https://hexdocs.pm/gql) ![CI](https://github.com/balexand/gql/actions/workflows/elixir.yml/badge.svg)

Simple GraphQL client for Elixir.

## Installation

The package can be installed by adding `gql` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:gql, "~> 0.6.0"}
  ]
end
```

## Usage

```elixir
url = "https://api.spacex.land/graphql/"

# List SpaceX launches
GQL.query("query Launches { launches { id } }", url: url)

# Get SpaceX launch by ID
GQL.query("query Launch($launch_id: ID!) { launch(id: $launch_id) { details } }",
  url: url,
  variables: [launch_id: "9"]
)
```

See [`GQL` docs](https://hexdocs.pm/gql/GQL.html) for details.

## Fragments

This library does not support GraphQL fragments. I'm open to pull-requests but this is not something that is a priority to me.

## Alternatives

Please submit an issue or pull-request if you know of any other GraphQL client libraries for Elixir.

### [`absinthe_client`](https://github.com/absinthe-graphql/absinthe_client)

The README indicates this project is not ready for use and it hasn't seen commits since 2019.

### [`neuron`](https://github.com/uesteibar/neuron)

Neuron is a great library and I have successfully used it in production. But there are a few issues that have lead me to create GQL instead:

* There are serious problems in how fragments are handled in `Neuron.Fragment`. All fragments are registered either globally using `Application.put_env/3` or per-process using `Process.put/2`. It would be easy for two different client libraries to overwrite each others' fragments and cause surprising and hard to find bugs.
* Neuron does not always return descriptive errors. For example, when an API endpoint returns an HTTP status of 500 with HTML in the body (like Shopify) then `Neuron.JSONParseError` is returned as opposed to an error that indicates 500 error. This makes the errors more cryptic and harder for apps like sentry.io to categorize.
* I prefer [`finch`](https://github.com/sneako/finch) over [`httpoison`](https://github.com/edgurgel/httpoison).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.6.2] - 2023-03-11
### Changed
- Relax `nimble_options` version requirement


test/gql_test.exs
defmodule GQLTest do
  use ExUnit.Case, async: true
  doctest GQL

  alias GQL.MockFinch

  @url "https://example.com/graphql"

  @spacex_launch_query """
  query Launch($launch_id: ID!) {
    launch(id: $launch_id) {
      id
      details
    }
  }
  """

  @headers [{"content-type", "application/json; charset=utf-8"}]

  @spacex_resp %Finch.Response{
    body:
      "{\"data\":{\"launches\":[{\"id\":\"13\",\"details\":\"Second GTO launch for Falcon 9. The USAF evaluated launch data from this flight as part of a separate certification program for SpaceX to qualify to fly U.S. military payloads and found that the Thaicom 6 launch had \\\"unacceptable fuel reserves at engine cutoff of the stage 2 second burnoff\\\"\"},{\"id\":\"17\",\"details\":null}]}}\n",
    headers: @headers,
    status: 200
  }

  @spacex_error_resp %Finch.Response{
    body:
      "{\"errors\":[{\"message\":\"Variable \\\"$launch_id\\\" of required type \\\"ID!\\\" was not provided.\",\"locations\":[{\"line\":1,\"column\":14}],\"extensions\":{\"code\":\"INTERNAL_SERVER_ERROR\"}}]}\n",
    headers: [{"content-type", "application/json; charset=utf-8"}],
    status: 400
  }

  @expected_body %{
    "data" => %{
      "launches" => [
        %{
          "details" =>
            "Second GTO launch for Falcon 9. The USAF evaluated launch data from this flight as part of a separate certification program for SpaceX to qualify to fly U.S. military payloads and found that the Thaicom 6 launch had \"unacceptable fuel reserves at engine cutoff of the stage 2 second burnoff\"",
          "id" => "13"
        },
        %{"details" => nil, "id" => "17"}
      ]
    }
  }

  import Mox, only: [verify_on_exit!: 1]
  setup :verify_on_exit!

  test "query" do
    Mox.expect(MockFinch, :request, fn req, GQL.Finch, [receive_timeout: 30000] ->
      assert req == %Finch.Request{
               body:
                 "{\"query\":\"query Launch($launch_id: ID!) {\\n  launch(id: $launch_id) {\\n    id\\n    details\\n  }\\n}\\n\",\"variables\":{}}",
               headers: [{"content-type", "application/json"}, {"X-Shopify-Access-Token", "🐓"}],
               host: "example.com",
               method: "POST",
               path: "/graphql",
               port: 443,
               query: nil,
               scheme: :https
             }

      {:ok, @spacex_resp}
    end)

    assert {:ok, @expected_body, @headers} ==
             GQL.query(@spacex_launch_query,
               finch_mod: MockFinch,
               headers: [{"X-Shopify-Access-Token", "🐓"}],
               url: @url
             )
  end

  test "query with GraphQL error" do
    Mox.expect(MockFinch, :request, fn _, _, _ ->
      {:ok, @spacex_error_resp}
    end)

    assert {:error, %{"errors" => [%{}]}, [_ | _]} =
             GQL.query(@spacex_launch_query, finch_mod: MockFinch, url: @url)
  end

  test "query without required opts" do
    assert_raise NimbleOptions.ValidationError,
                 "required :url option not found, received options: [:variables, :http_options, :headers, :finch_mod]",
                 fn ->
                   GQL.query(@spacex_launch_query, [])
                 end
  end

  test "query with HTTP 500 error" do
    Mox.expect(MockFinch, :request, fn _, _, _ ->
      {:ok, %Finch.Response{body: "boom", headers: [], status: 500}}
    end)

    assert_raise GQL.ServerError, "Server responded with HTTP status: 500", fn ->
      GQL.query(@spacex_launch_query, finch_mod: MockFinch, url: @url)
    end
  end

  test "query!" do
    Mox.expect(MockFinch, :request, fn _, _, _ ->
      {:ok, @spacex_resp}
    end)

    assert {@expected_body, @headers} ==
             GQL.query!(@spacex_launch_query, finch_mod: MockFinch, url: @url)
  end

  test "query! with GraphQL error" do
    Mox.expect(MockFinch, :request, fn _, _, _ ->
      {:ok, @spacex_error_resp}
    end)

    assert_raise GQL.GraphQLError,
                 "%GQL.GraphQLError{body: %{\"errors\" => [%{\"extensions\" => %{\"code\" => \"INTERNAL_SERVER_ERROR\"}, \"locations\" => [%{\"column\" => 14, \"line\" => 1}], \"message\" => \"Variable \\\"$launch_id\\\" of required type \\\"ID!\\\" was not provided.\"}]}}",
                 fn ->
                   GQL.query!(@spacex_launch_query, finch_mod: MockFinch, url: @url)
                 end
  end
end


test/test_helper.exs
ExUnit.configure(exclude: [integration: true])

ExUnit.start()


test/support/mocks.ex
defmodule GQL.Finch.Behaviour do
  @callback request(Finch.Request.t(), Finch.name(), keyword()) ::
              {:ok, Finch.Response.t()} | {:error, Finch.Exception.t()}
end

Mox.defmock(GQL.MockFinch, for: GQL.Finch.Behaviour)


mix.exs
defmodule GQL.MixProject do
  use Mix.Project

  @version "0.6.2"

  def project do
    [
      app: :gql,
      description: "Simple GraphQL client.",
      version: @version,
      elixir: "~> 1.12",
      elixirc_paths: elixirc_paths(Mix.env()),
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      package: [
        licenses: ["MIT"],
        links: %{"GitHub" => "https://github.com/balexand/gql"}
      ],
      docs: [
        extras: ["README.md"],
        source_ref: "v#{@version}",
        source_url: "https://github.com/balexand/gql"
      ]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger],
      mod: {GQL.Application, []}
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:finch, "~> 0.12"},
      {:jason, "~> 1.3"},
      {:nimble_options, "~> 0.4 or ~> 1.0"},

      # Dev/test
      {:dialyxir, "~> 1.0", only: [:dev], runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:mox, "~> 1.0", only: :test}
    ]
  end
end


lib/gql.ex
defmodule GQL do
  @moduledoc """
  Simple GraphQL client.
  """

  defmodule Behaviour do
    @moduledoc """
    Behaviour that `GQL` implements.
    """

    @callback query!(String.t(), keyword()) :: {map(), Mint.Types.headers()}
    @callback query(String.t(), keyword()) ::
                {:ok, map(), Mint.Types.headers()} | {:error, map, Mint.Types.headers()}
  end

  @behaviour Behaviour

  defmodule ConnectionError do
    @moduledoc """
    Error raised when a connection error occurs. See `Mint.TransportError` for list of possible
    values for the `reason` field.
    """

    defexception [:reason]

    def message(exception) do
      inspect(exception)
    end
  end

  defmodule GraphQLError do
    @moduledoc """
    Error raised when response contains GraphQL errors.
    """

    defexception [:body]

    def message(exception), do: inspect(exception)
  end

  defmodule ServerError do
    @moduledoc """
    Error raised when server returns 5xx HTTP status code.
    """

    defexception [:response, :status]

    def message(exception), do: "Server responded with HTTP status: #{exception.status}"
  end

  @query_opts_validation [
    finch_mod: [
      type: :atom,
      default: Finch,
      doc: false
    ],
    headers: [
      type: {:list, :any},
      default: [],
      doc: "HTTP headers to include."
    ],
    http_options: [
      type: :keyword_list,
      doc: "Options to be passed to `Finch.request/3`.",
      default: [receive_timeout: 30_000]
    ],
    variables: [
      type: :any,
      default: [],
      doc: "Keyword list or map of variables."
    ],
    url: [
      type: :string,
      required: true,
      doc: "URL to which the request is made."
    ]
  ]

  @doc """
  Like `query/2`, except raises `GQL.GraphQLError` if the server returns errors.
  """
  @impl true
  @spec query!(String.t(), keyword()) :: {map(), Mint.Types.headers()}
  def query!(query, opts) do
    case query(query, opts) do
      {:ok, body, headers} -> {body, headers}
      {:error, body, _headers} -> raise %GraphQLError{body: body}
    end
  end

  @doc """
  Queries a GraphQL endpoint. Returns `{:ok, body, headers}` upon success or `{:error, body,
  headers}` if the response contains an "errors" key.

  An exception will be raised for exceptional errors:

  * `GQL.ConnectionError` if the HTTP client returns a connection error such as a timeout.
  * `GQL.ServerError` if the server responded with a 5xx code.

  ## Options

  #{NimbleOptions.docs(@query_opts_validation)}
  """
  @impl true
  @spec query(String.t(), keyword()) ::
          {:ok, map(), Mint.Types.headers()} | {:error, map, Mint.Types.headers()}
  def query(query, opts) do
    opts = NimbleOptions.validate!(opts, @query_opts_validation)

    body = %{query: query, variables: Map.new(opts[:variables])}
    headers = [{"content-type", "application/json"}] ++ opts[:headers]

    Finch.build(:post, opts[:url], headers, Jason.encode!(body))
    |> opts[:finch_mod].request(GQL.Finch, opts[:http_options])
    |> case do
      {:ok, %Finch.Response{status: status} = resp} when status >= 200 and status < 500 ->
        handle_body(Jason.decode!(resp.body), resp.headers)

      {:ok, %Finch.Response{} = resp} ->
        raise %ServerError{response: resp, status: resp.status}

      {:error, %Mint.TransportError{reason: reason}} ->
        raise %ConnectionError{reason: reason}
    end
  end

  defp handle_body(%{"errors" => _} = body, headers) do
    # Return error if response body contains errors. In this case the HTTP status is inconsistent
    # between different APIs. Github and SpaceX return errors with HTTP 200 status. Shopify
    # returns errors with HTTP 400 status.
    {:error, body, headers}
  end

  defp handle_body(%{} = body, headers) do
    {:ok, body, headers}
  end
end
